metaflow/flowspec.py

```diff
@@ -76,14 +77,84 @@ def __getitem__(self, item):
         return item or 0  # item is None for the control task, but it is also split 0
 
 
-class _FlowState(Enum):
+class FlowStateItems(Enum):
     CONFIGS = 1
     FLOW_MUTATORS = 2
     CACHED_PARAMETERS = 3
-    SET_CONFIG_PARAMETERS = (
-        4  # These are Parameters that now have a ConfigValue (converted)
-    )
-    # but we need to remember them.
+    SET_CONFIG_PARAMETERS = 4  # Parameters that now have a ConfigValue (converted)
+    FLOW_DECORATORS = 5
+
+
+class _FlowState(MutableMapping):
+    # Dict like structure to hold state information about the flow but it holds
+    # the key/values in two sub dictionaries: the ones that are specific to the flow
+    # and the ones that are inherited from parent classes.
+    # This is NOT a general purpose class and is meant to only work with FlowSpec.
+    # For example, it assumes that items are only list, dicts or None and assumes that
+    # self._self_data has all keys properly initialized.
+
+    def __init__(self, *args, **kwargs):
+        self._self_data = dict(*args, **kwargs)
+        self._merged_data = {}
+        self._inherited = {}
+
+    def __getitem__(self, key):
+        # ORDER IS IMPORTANT: we use inherited first and extend by whatever is in
+        # the flowspec
+        if key in self._merged_data:
+            return self._merged_data[key]
+
+        # We haven't accessed this yet so compute it for the first time
+        self_value = self._self_data.get(key)
+        inherited_value = self._inherited.get(key)
+
+        if self_value is not None:
+            self._merged_data[key] = self._merge_value(inherited_value, self_value)
+            return self._merged_data[key]
+        elif key in self._self_data:
+            # Case of CACHED_PARAMETERS; a valid value is None. It is never inherited
+            self._merged_data[key] = None
+            return None
+        raise KeyError(key)
```

Comment:
The root cause of this bug is here . This bug is triggered because `self._merged_data` can be set when the user does something like in L545 (ie doing a getter like `cls._flow_state[_FlowState.SET_CONFIG_PARAMETERS]` ). 

This pattern is a little funky because based on how `__getitem__` is written, if a key is accessed once then it will be set in the `self._merge_data` and once it is set in the `self._merge_data` then even if we call `__setitem__`, the `__getitem__` will ONLY return the older value since it's in `self._merge_data`